### PR TITLE
Transformed all remaining occurences of for..of to forEach()

### PR DESCRIPTION
**Pull request by rhettlivingston**

### CODE REVIEW

**Comment by benmosher** [comment]
fixes #415?

File: src/core/getExports.js [hunk 1]
```diff
@@ -236,18 +236,19 @@ export default class ExportMap {
     if (this.reexports.has(name)) return true
 
     // default exports must be explicitly re-exported (#328)
+    let foundInnerMapName = false
     if (name !== 'default') {
-      for (let dep of this.dependencies.values()) {
-        let innerMap = dep()
+      this.dependencies.forEach((dep) => {
```
Comment (benmosher) [inline src/core/getExports.js]: this would ideally be a `some` so that it short-circuits. not sure how to achieve with a `Set`...

**Comment by rhettlivingston** [comment]
Yes. Didn't put that in the comment because I was merging into the branch
that fixes it.

I have to apologize for what it does to your code. I really wish I knew how
to just make the for..of implementation play well with others. forEach is
very inferior in this case because you can neither 'continue' within it or
'return' from it.

File: src/core/getExports.js [hunk 2]
```diff
@@ -276,24 +277,29 @@ export default class ExportMap {
...
-        if (!innerMap) continue
-
-        // safeguard against cycles
-        if (innerMap.path === this.path) continue
-
-        let innerValue = innerMap.hasDeep(name)
-        if (innerValue.found) {
-          innerValue.path.unshift(this)
-          return innerValue
+      this.dependencies.forEach((dep) => {
```
Comment (benmosher) [inline src/core/getExports.js]: same here, ideally would short-circuit

File: src/rules/no-unresolved.js [hunk 3]
```diff
@@ -53,15 +53,17 @@ module.exports = function (context) {
...
-    for (let element of modules.elements) {
-      if (element.type !== 'Literal') continue
-      if (typeof element.value !== 'string') continue
-
-      if (element.value === 'require' ||
-          element.value === 'exports') continue // magic modules: http://git.io/vByan
-
-      checkSourceValue(element)
-    }
+    modules.elements.forEach((element) => {
```
Comment (benmosher) [inline src/rules/no-unresolved.js]: I would prefer that these `continue`s become `returns`, to minimize deep nesting. but that's just code style preference, no impact on semantics/performance 🤓

File: src/core/getExports.js (hunk 1 above)
Comment (rhettlivingston) [inline src/core/getExports.js]: I learned while doing this that `Maps` and `Sets` are very fundamentally not `Arrays`. So, no `some` and not even a clear way to utilize the basic `for` statement and break out of it (i.e. you can't access them in an indexed fashion).

I suspect the overhead of converting to an array just to use some would be greater than the gain.

File: src/rules/no-unresolved.js (hunk 3 above)
Comment (rhettlivingston) [inline src/rules/no-unresolved.js]: Ahhh. Realized that returns didn't do the same thing inside the foreach, but didn't click on the fact that they did do the same thing as a continue. I agree with you. Will go back and take advantage of that insight. It twists your code a little less.

Frankly, I prefer your original code style and have used it in many similar cases for more than 30 years despite the CompSci perfectionists always criticizing me for not having a single point of return at the end of my functions. I'm a CompEng. We prefer to do what works. :smile:

File: src/core/getExports.js (hunk 1 above)
Comment (rhettlivingston) [inline src/core/getExports.js]: After thinking about this some,,, I may experiment with directly accessing the iterator in a for loop. This is essentially what is going on with for..of, but I might be able to get away with it if I don't involve the underlying Symbol type and go directly to the Map / Set.

I've also realized I might have changed more than I had too. The instances that definitely caused problems involved using for..of to iterate Maps. In response, I changed all occurrences of for..of. It is possible that using for..of to iterate a Set causes no issue.

**Comment by benmosher** [comment]
**On short-circuits**: I think you're probably right, best to leave it be.

Though notably, `this.dependencies` only ever deals with its values after parse time, so the only use it gets being a `Map` is one-per-path `export * from x` behavior, which could be easily achieved with an ephemeral `Set` of paths during `parse` and just `if (!exportAllPaths.has(remoteMap)) m.dependencies.push(() => ExportMap.for(remoteMap, context)` on line ~163 of `getExports.js`.

FWIW, I'll probably revert to the `for ... of` syntax in v2 anyway since 0.10 will no longer be supported. so no sweat either way. 😁 Best to just get it correct for the short term and not worry too much about the ideal case.

I'm just over-thinking it.

**Comment by benmosher** [comment]
I'll go ahead and merge as-is if it fixes #415, no need to sweat the style stuff ATM.

**Comment by rhettlivingston** [comment]
Thank you very much for accepting this. And I'd absolutely agree with reverting to `for..of` later. It is unquestionably the proper way to traverse Maps and Sets.

**Pull request merged**